imu.c

```diff
@@ -0,0 +1,534 @@
+
+
+uint8_t indx = 0;
+float accX = 0, accY = 0, accZ = 0;
+float gyrX = 0, gyrY = 0, gyrZ= 0;
+struct bmi2_sens_config config;
+
+// bytes to be sent over SPI
+uint8_t imuSPIBuffer[20];
+
+
+int8_t imuRead(uint8_t reg_addr, uint8_t *reg_data, uint32_t len, void *intf_ptr){
+        
+        imuSPIBuffer[0] = reg_addr;
+        memcpy(imuSPIBuffer+1, reg_data, len);
+
+        spiTransfer(SPI_BUS, imuSPIBuffer, len + 1);
+
+        memcpy(reg_data,imuSPIBuffer+1,len);
+
+        return BMI2_INTF_RET_SUCCESS;
+
+}
+
+int8_t imuWrite(uint8_t reg_addr, const uint8_t *reg_data, uint32_t len, void *intf_ptr){
+
+    imuSPIBuffer[0] = reg_addr;
+    memcpy(imuSPIBuffer+1, reg_data, len);
+    spiTransfer(SPI_BUS,imuSPIBuffer,len+1);
+
+    //Debugging
+    /*
+    if(len == 1){
+        NRF_LOG_RAW_INFO("len = %d (raw) reg_addr = %x, data = %x\n",len,  reg_addr, *reg_data);
+    }
+    else if(len == 2){
+        NRF_LOG_RAW_INFO("len = %d (raw) reg_addr = %x, data = %x %x \n", len,  reg_addr, *reg_data, *(reg_data + 1));
+    }
+    else{
+        NRF_LOG_RAW_INFO("len = %d (raw) reg_addr = %x, data = %x %x %x \n", len,  reg_addr, *reg_data, *(reg_data + 1), *(reg_data + 2));
+    }
+    */
+
+    return BMI2_INTF_RET_SUCCESS;
+}
+
+void imuDelay(uint32_t period, void *intf_ptr){
+    //delayMs(period);
+    nrf_delay_us(period);
+}
+
+void imuInit(void)
+{
+    // Initialize SPI interface
+    //BMI270_sensor_init();
+    rslt = bmi270_init(&bmi);
+    bmi2_print_error_code(rslt);
+
+    if(rslt == BMI2_OK)
+    {
+        // Accel configuration settings. 
+        rslt = set_accel_gyro_config(&bmi);
+        bmi2_print_error_code(rslt);
+
+        // NOTE: Accel enable must be done after setting configurations
+        rslt = bmi2_sensor_enable(sensor_list, 2, &bmi);
+        bmi2_print_error_code(rslt);
+
+        NRF_LOG_RAW_INFO("IMU initialization successful\n")
+    }
+    else{
+
+        NRF_LOG_RAW_INFO("IMU initialization failed\n")
+    }
+
+
+}
+
+void imuReadAccel(void){
+
+    config.type = BMI2_ACCEL;
+
+    // Get the accel configurations. 
+    rslt = bmi2_get_sensor_config(&config, 1, &bmi);
+    bmi2_print_error_code(rslt);
+
+    rslt = bmi2_get_sensor_data(&sensor_data, &bmi);
+    bmi2_print_error_code(rslt);
+
+    if ((rslt == BMI2_OK) && (sensor_data.status & BMI2_DRDY_ACC))
+    {
+        //Converting lsb to meter per second squared for 16 bit accelerometer at 2G range.
+        accX = lsb_to_mps2(sensor_data.acc.x, (float)2, bmi.resolution);
+        accY = lsb_to_mps2(sensor_data.acc.y, (float)2, bmi.resolution);
+        accZ = lsb_to_mps2(sensor_data.acc.z, (float)2, bmi.resolution);
+        NRF_LOG_RAW_INFO("Accel: x y z \n");
+        NRF_LOG_RAW_INFO(NRF_LOG_FLOAT_MARKER "  ", NRF_LOG_FLOAT(accX));
+        NRF_LOG_RAW_INFO(NRF_LOG_FLOAT_MARKER "  ", NRF_LOG_FLOAT(accY));
+        NRF_LOG_RAW_INFO(NRF_LOG_FLOAT_MARKER"\n",NRF_LOG_FLOAT(accZ));
+
+    }
+
+}
+
+void imuReadGyro(void){
```

Comment:
we're going to want to pass the imuData over to another caller. Either make this accept a uint16_t* pointer, or create readAccelX, readAccelY, ... and have them return uint16_t. We can do the sensor conversion on the phone and just send the raw bytes.